### PR TITLE
[SIMPLY-2580] Publish the 'BookOpened' event from both R1 and R2 readers.

### PR DESCRIPTION
**Pull request by twaddington**
**What's this do?**
Publish the 'BookOpened' event from both R1 and R2 readers.

**Why are we doing this? (w/ JIRA link if applicable)**
https://jira.nypl.org/browse/SIMPLY-2580

**How should this be tested? / Do these changes have associated tests?**
n/a

**Dependencies for merging? Releasing to production?**
We'll need to add `:simplified-analytics-circulation` as a dependency to SimplyE to properly consume these events.

**Has the application documentation been updated for these changes?**
n/a

**Did someone actually run this code to verify it works?**
@twaddington ran the Vanilla app.

### CODE REVIEW

File: simplified-analytics-circulation/src/main/java/org/nypl/simplified/analytics/circulation/CirculationAnalyticsSystem.kt [hunk 1]
```diff
@@ -25,47 +26,23 @@ class CirculationAnalyticsSystem(
...
-          this.logger.debug(
-            "no analytics URI available for book {} ({})",
-            event.opdsEntry.id,
-            event.opdsEntry.title)
+        event.targetURI?.let { target ->
+          postURI(target, event.credentials)
         }
       }
+      else -> {
+        // All other events are silently dropped
```
Comment (twaddington) [inline simplified-analytics-circulation/src/main/java/org/nypl/simplified/analytics/circulation/CirculationAnalyticsSystem.kt:37]: @io7m are all these other events are used by LFA?

File: simplified-viewer-epub-readium2/src/main/java/org/nypl/simplified/viewer/epub/readium2/ReaderActivity.kt [hunk 2]
```diff
@@ -149,6 +164,19 @@ class ReaderActivity : AppCompatActivity(), SR2ControllerHostType {
...
+      this.analyticsService.publishEvent(
+        AnalyticsEvent.BookOpened(
+          timestamp = LocalDateTime.now(),
+          credentials = this.account.loginState.credentials,
+          profileUUID = this.currentProfile.id.uuid,
+          profileDisplayName = this.currentProfile.displayName,
+          accountProvider = this.account.provider.id,
+          accountUUID = this.account.id.uuid,
+          opdsEntry = this.bookEntry.feedEntry,
+          targetURI = this.bookEntry.feedEntry.analytics.getOrNull()
```
Comment (twaddington) [inline simplified-viewer-epub-readium2/src/main/java/org/nypl/simplified/viewer/epub/readium2/ReaderActivity.kt:175]: I'm pretty sure this is the right value to set here but I saw some funky results for some books. @io7m any ideas?

File: simplified-viewer-epub-readium2/src/main/java/org/nypl/simplified/viewer/epub/readium2/ReaderActivity.kt [hunk 3]
```diff
@@ -149,6 +164,19 @@ class ReaderActivity : AppCompatActivity(), SR2ControllerHostType {
         .beginTransaction()
         .replace(R.id.reader_container, readerFragment)
         .commit()
+
+      this.analyticsService.publishEvent(
+        AnalyticsEvent.BookOpened(
+          timestamp = LocalDateTime.now(),
+          credentials = this.account.loginState.credentials,
+          profileUUID = this.currentProfile.id.uuid,
```
Comment (twaddington) [inline simplified-viewer-epub-readium2/src/main/java/org/nypl/simplified/viewer/epub/readium2/ReaderActivity.kt:170]: None of these properties seem to be read. Maybe they're used by LFA?

**Comment by io7m** [comment]
I have to rain on the parade a little. I think it might be better to implement this here:

https://github.com/NYPL-Simplified/Simplified-Android-Core/blob/develop/simplified-viewer-api/src/main/java/org/nypl/simplified/viewer/api/Viewers.kt#L31

Publish one event from the code that opens the viewer, rather than one event from each type of viewer. We did this once and then I accidentally removed it (leading to you having to write this PR in the first place). :sweat:

**Comment by twaddington** [comment]
> I have to rain on the parade a little. I think it might be better to implement this here:
> 
> https://github.com/NYPL-Simplified/Simplified-Android-Core/blob/develop/simplified-viewer-api/src/main/java/org/nypl/simplified/viewer/api/Viewers.kt#L31
> 
> Publish one event from the code that opens the viewer, rather than one event from each type of viewer. We did this once and then I accidentally removed it (leading to you having to write this PR in the first place). 😓

That works for me. How should the close book event be handled then? I didn't want to remove that in case it was used by LFA.

**Comment by twaddington** [comment]
@io7m updated.

**Comment by io7m** [comment]
> That works for me. How should the close book event be handled then? I didn't want to remove that in case it was used by LFA.

That is a good question. As far as I can tell... Android makes it fundamentally impossible to handle reliably; it's inevitably going to be tied to a viewer activity closing, and we've no way to know if the activity closed because the user stopped reading, or it closed because that's what Android does.

**Comment by twaddington** [comment]
> > That works for me. How should the close book event be handled then? I didn't want to remove that in case it was used by LFA.
> 
> That is a good question. As far as I can tell... Android makes it fundamentally impossible to handle reliably; it's inevitably going to be tied to a viewer activity closing, and we've no way to know if the activity closed because the user stopped reading, or it closed because that's what Android does.

We could publish the event only from an intentional user interaction like `onBackPressed` or if the toolbar back is pressed.

**Pull request merged**